### ISSUE TITLE
Unable to reproduce results

### ISSUE THREAD

**Issue by FredHaa** [The-Learning-Machines/Audiomer-PyTorch#1]
Hi

Thank you for uploading your code.

In order to get it to work I had to change the ComposeMany transform to Compose, as ComposeMany adds a dimension to the waveform which makes the rest of the code fail. You don't generate multiple samples anyways, so no need to use ComposeMany.

Once that was fixed I tried running your experiments, and I fail to reproduce the results.
I ran run_expts.pywith changing any parameters (i.e. 64 batch size, 8192 sample rate, etc.) but changed it to only train the L model on SC12. Here I expected to reach 99.98% accuracy. While I did not actually try it on the test set, I only reached about 95% on the validation set. The training accuracy reached 99.8%.

Do you have any idea why I'm not seeing the same results as you? Maybe change the sample rate to 16k?

I'm using PyTorch 1.10 on a machine with a single RTX 3090.

**Comment by yodakohl** [comment]
I had a similar problem but fixed the additional dimension using torch.squeeze(). I trained the S model on SC12
and it seems like I got a good result. 

I also had to modify the evaluation code to run the test. 

```
diff --git a/evaluate.py b/evaluate.py
index b188e02..003603e 100644
--- a/evaluate.py
+++ b/evaluate.py
@@ -41,6 +41,10 @@ class Experiment(pl.LightningModule):
         self.loss_fn = single_label_cross_entropy
         self.activation_function = PartialModule(torch.log_softmax, dim=-1)
         self.model_params['pool'] = 'cls'
+        self.model_params['use_residual']=True
+        self.model_params['use_attention']=True
+        self.model_params['equal_strides']=True
+        self.model_params['use_se']=True
 
         # networks
         if kwargs['dataset'] == 'SC35':

```

The loss curve is attached. I'm running pytorch 1.9.0.

Testing: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊| 2035/2037 [00:36<00:00, 56.19it/s]--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'ACC/test': 0.9997545480728149, 'LOSS/test': 0.0015029070200398564}
------------------------------------------------------------------------------
![audiomer_sc12_s](https://user-images.githubusercontent.com/819019/140510023-928c8f22-dccf-4667-b5ab-d8a32bcbf99c.png)
--

**Comment by FredHaa** [comment]
I'll downgrade to 1.9.0 and try again.

**Comment by ojus1** [comment]
I am very sorry that I released untested code! We had to clean up alot of the stuff that isn't in the paper (but may come up as another paper). @yodakohl feel free to do a pull request, ill merge it. Thanks a lot!

**Comment by FredHaa** [comment]
Sorry I didn't reply for a second there, I had a problem with the work station where it constantly rebooted interrupting the training. Turns out it was the thermals which I've fixed now.

I've pulled the latest changes to fix the dimensionality bug and trained a new S network on SC12.

Here is my training/val loss:

![Screenshot from 2021-11-09 19-50-52](https://user-images.githubusercontent.com/6875946/140986319-74fb5812-f138-454a-a3df-5455e7585e32.png)

By looking at the validation loss it can clearly be seen that the model overfits on the training data.

Here is the output of the test evaluation:

--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'ACC/test': 0.9408443570137024, 'LOSS/test': 0.2121301293373108}
--------------------------------------------------------------------------------

Any ideas whats causing this?

**Comment by ojus1** [comment]
I have no idea. After this issue was opened, I updated all of the packages in requirements.txt, i am able to reproduce results in two different machines. 

I just don't get the same curves as yours.
Can you paste  your version of the packages by running `pigar` [link](https://github.com/Damnever/pigar)?

**Comment by FredHaa** [comment]
# Automatically generated by https://github.com/damnever/pigar.

# Audiomer-PyTorch/Audiomer/module.py: 6
# Audiomer-PyTorch/evaluate.py: 1
# Audiomer-PyTorch/experiments.py: 1
einops == 0.3.0

# Audiomer-PyTorch/Metrics.py: 4
# Audiomer-PyTorch/evaluate.py: 7
# Audiomer-PyTorch/experiments.py: 7
numpy == 1.19.5

# Audiomer-PyTorch/Audiomer/module.py: 1,2
performer_pytorch == 1.1.2

# Audiomer-PyTorch/datamodules/SpeechCommands12.py: 14
# Audiomer-PyTorch/datamodules/SpeechCommands20.py: 14
# Audiomer-PyTorch/datamodules/SpeechCommands35.py: 14
# Audiomer-PyTorch/evaluate.py: 5,193,194
# Audiomer-PyTorch/experiments.py: 5,170,171
pytorch_lightning == 1.5.0

# Audiomer-PyTorch/Audiomer/module.py: 3,4,5
# Audiomer-PyTorch/Audiomer/utils.py: 1,2,3
# Audiomer-PyTorch/Metrics.py: 2
# Audiomer-PyTorch/datamodules/SpeechCommands12.py: 16,19
# Audiomer-PyTorch/datamodules/SpeechCommands20.py: 16,19
# Audiomer-PyTorch/datamodules/SpeechCommands35.py: 16,19
# Audiomer-PyTorch/datamodules/pad.py: 1
# Audiomer-PyTorch/evaluate.py: 8,9
# Audiomer-PyTorch/experiments.py: 8,9
torch == 1.9.0+cu111

# Audiomer-PyTorch/datamodules/SpeechCommands12.py: 17,18
# Audiomer-PyTorch/datamodules/SpeechCommands20.py: 17,18
# Audiomer-PyTorch/datamodules/SpeechCommands35.py: 17,18
torchaudio == 0.9.0

# Audiomer-PyTorch/datamodules/SpeechCommands12.py: 1
# Audiomer-PyTorch/datamodules/SpeechCommands20.py: 1
# Audiomer-PyTorch/datamodules/SpeechCommands35.py: 1
torchaudio_augmentations == 0.2.2

# Audiomer-PyTorch/Metrics.py: 1
torchmetrics == 0.6.0

# Audiomer-PyTorch/Audiomer/module.py: 238
torchsummary == 1.5.1

**Comment by ojus1** [comment]
Here's mine:
einops == 0.3.2
numpy == 1.21.2
performer_pytorch == 1.1.3
pytorch_lightning == 1.5.0
torch == 1.10.0
torchaudio == 0.10.0
torchaudio_augmentations == 0.2.2
torchmetrics == 0.6.0
torchsummary == 1.5.1

I can't think of anything else that might be affecting the results.

**Comment by FredHaa** [comment]
I'll try your versions and report back.

**Comment by FredHaa** [comment]
Still no dice.

Have you tried to train a network with this codebase as is? I.e. after you cleaned up the repo?
I'll try to spin up a colab session to see if it is somehow my machine which does something funky.

**Comment by ojus1** [comment]
Yes, I've trained two models on two different machines and my results match the reported numbers. I used the exact same repository. Maybe try training on a different device (not on your current gpu)?

**Comment by FredHaa** [comment]
I'm currently training on my home desktop and it seems like it will reach the accuracy reported in the paper, no sign of overfitting!
I'm using a GTX 1060 so it takes forever.

I'm going to reinstall the RTX 3090 workstation with Ubuntu 20.04 (it currently runs 21.04) as well as CUDA 11.3 (it runs 11.1) in order to mirror the setup on my desktop and hopefully see that I can obtain similar results.

But no matter what, it doesn't seem to be an error with the code, more likely a bug in the specific combination of hardware and software.

**Comment by ojus1** [comment]
This is absurd, I would think it would be related to CUDA, since alot of operations in cuda is non-deterministic. I had heard about problems like these, however, it was my first time hearing about it first hand.
All the best to you!
